Safepoints lazy fix -- Attempt 2

kayceesrk opened the pull request:
This supplants #492 and #493. This also addresses comments in https://github.com/ocaml-multicore/ocaml-multicore/pull/492#pullrequestreview-612034443.

* Adds the test from #483 as `lazy7.ml`
* The PR moves more of the internal implementation of lazy into the runtime so that we don't make assumptions about atomicity w.r.t the GC. 

## Notes for reviewing

The first commit does the bulk of the work. It may be useful to review that commit first.

kayceesrk committed:
Make the implementation of lazy aware of safe points.

Since safe points may be inserted by the compiler at the start of
functions and back edges of loops, more lazy functionality must be
implemented in the runtime.

kayceesrk committed:
Remove the use of make_forward from camlinternalMod

kayceesrk committed:
Merge branch '4.12+domains+effects' into safepoints_lazy_fix2

kayceesrk committed:
Remove unused primitives. Requires boostrap.

kayceesrk committed:
Commit bootstrap results

kayceesrk committed:
Remove the incorrect assertions

kayceesrk committed:
Spawn fewer domains to keep CI happy

kayceesrk committed:
Added lazy8.ml to test the refined behaviour of Lazy.force_val.

The new behaviour is a refinement over the stock behaviour, and does not
conflict with stock's behaviour with lazy. In particular, the new
behaviour for sequential code is a subset of the currently observable
behaviour.

When Lazy.force_val raises an exception, stock OCaml says that force_val
on the same lazy again may raise either Undefined or the original
exception. The new behaviour is that, if the subsequent force_val of the
lazy was done on the same domain, then Undefined exception is raised.
Otherwise, RacyLazy exception is raised.

Sudha247 commented on stdlib/camlinternalLazy.ml:
```diff
@@ -50,43 +83,37 @@ let do_force_val_block blk =
-  | Forcing when only_val -> Some (do_force_val_block blk)
-  | Forcing -> Some (do_force_block blk)
+let try_force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
+  match update_to_forcing (Obj.repr blk) with
```
I suspect it's possible to run into the same error we were previously witnessing here.

Sudha247 commented on runtime/obj.c:
```diff
@@ -250,6 +212,74 @@ CAMLprim value caml_lazy_make_forward (value v)
+  tag_t tag;
+  value field0;
+
+  if (obj_update_tag (v, Lazy_tag, Forcing_tag)) {
```
I think we should be making sure `v` is still a block before we pass it to `obj_update_tag`. In the case that `v` is not a block, this could lead to the same error we were witnessing previously. We should probably go over the other checks first before calling `obj_update_tag`.

kayceesrk replied:
That's correct. Good catch.

Sudha247 reviewed (commented):
Thank you, this is a great improvement over the existing implementation! Moving bits to the runtime is helpful. 

I just had one suggestion on `update_to_forcing`.

kayceesrk committed:
Fix bug in caml_lazy_update_to_forcing.

The function now handles the case where the object may have been
short-circuited. We also enforce ordering constraints in the function by
using an atomic load instruction.

Also document the atomicity expectations w.r.t GC in
camlinternalLazy.ml.

kayceesrk committed:
Merge upstream

kayceesrk commented:
@Sudha247 I've fixed the bug. I've also made other improvements in https://github.com/ocaml-multicore/ocaml-multicore/pull/501/commits/8217553e5934feff38e90b63faea52815f2cf221. Please review.

Sudha247 commented on stdlib/camlinternalLazy.ml:
```diff
@@ -50,58 +83,56 @@ let do_force_val_block blk =
    for debugging purpose. *)
 let force_gen ~only_val (lzv : 'arg lazy_t) =
   let x = Obj.repr lzv in
+  (* XXX No safe points start *)
```
Just so I'm not misreading this comment: Does this mean we are assuming there are no safe points in this range or that the compiler will not insert any safe points?

kayceesrk replied:
We are assuming that the compiler will not insert safe points in this block. The compiler only inserts safe points at function entries and back edges of loops now. If this behaviour were to change, then this code will break. This is documenting the assumptions that hold now.

Sudha247 reviewed (approved):
Great! This fixes the bug in #483. The changes look good to me.

I had a question on a comment, but that shouldn't block merging this.
